File: c-kernel/test/test_params.py
import os
import sys

sys.path.append('../../arl-python')

import time
import numpy as np
import argparse

from arl.imaging.params import get_rowmap
from utils import store_data

def test_params(data_dir, nrows, low, high, len):
    frequency = np.random.choice(np.linspace(low, high, len), nrows)
    ufrequency = np.unique(frequency)

    store_data(os.path.join(data_dir, 'frequency.dat'), frequency)
    store_data(os.path.join(data_dir, 'ufrequency.dat'), ufrequency)

    start = time.time()
    vmap = np.array(get_rowmap(frequency, ufrequency))
    stop = time.time()
    print('Original Time:  {:.2f}s'.format(stop - start))

    store_data(os.path.join(data_dir, 'vmap.dat'), vmap)


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--data_dir', type=str, default='./data')
    parser.add_argument('--nrows', type=int, default=0)
    parser.add_argument('--low', type=float, default=0)
    parser.add_argument('--high', type=float, default=0)
    parser.add_argument('--len', type=int, default=0)
    args = parser.parse_args()

    test_params(args.data_dir, args.nrows, args.low, args.high, args.len)

File: c-kernel/test/test_params.cc
// test/test_params.cc
// tests of parameters
// Author: You Haihang, Yang Runkai, Liu Tao,
// Applied Computing Lab, ICT, CAS

#include <iostream>
#include <string>
#include <ctime>
#include <getopt.h>

#include "params.h"
#include "utils.h"

using namespace std;

void test_params(string &data_dir, int nrows, int n_unique) {
    double *frequency = new double[nrows];
    double *ufrequency = new double[n_unique];
    int *vmap = new int[nrows];
    int *vmap_exact = new int[nrows];

    load_data(data_dir + "/" + "frequency.dat", frequency, nrows);
    load_data(data_dir + "/" + "ufrequency.dat", ufrequency, n_unique);
    load_data(data_dir + "/" + "vmap.dat", vmap_exact, nrows);

    clock_t start = clock();
    get_rowmap(vmap, frequency, ufrequency, nrows, n_unique);
    clock_t stop = clock();
    printf("Optimized Time: %.2lfs\n", 1.0 * (stop - start) / CLOCKS_PER_SEC);

    double err = diff(vmap, vmap_exact, nrows);
    cout << "Params Error: " << err << endl;

    delete [] frequency;
    delete [] ufrequency;
    delete [] vmap_exact;
    delete [] vmap;
}

int main(int argc, char *argv[]) {
    string data_dir = "./data";
    int nrows = 0;
    double low = 0.0;
    double high = 0.0;
    int len = 0;

    int c;
    while (true) {
        int option_index = 0;
        static struct option long_options[] = {
            {"data_dir", required_argument, 0, 1},
            {"nrows", required_argument, 0, 2},
            {"low", required_argument, 0, 3},
            {"high", required_argument, 0, 4},
            {"len", required_argument, 0, 5}
        };

        c = getopt_long(argc, argv, "", long_options, &option_index);
        if (c == -1) break;

        switch (c) {
            case 1: data_dir = string(optarg); break;
            case 2: nrows = atoi(optarg); break;
            case 3: low = atof(optarg); break;
            case 4: high = atof(optarg); break;
            case 5: len = atoi(optarg); break;
            default: cout << "Wrong options." << endl; break;
        }
    }

    test_params(data_dir, nrows, len);
}
